Stack/Opc.Ua.Core/Security/Certificates/CertificateFactory.cs

```diff
@@ -84,10 +103,337 @@ public static X509Certificate2 Load(X509Certificate2 certificate, bool ensurePri
+
+            // Basic constraints
+            cg.AddExtension(X509Extensions.BasicConstraints.Id, true, new BasicConstraints(isCA));
+
+            // Authority Key identifier
+            var issuerKeyPair = subjectKeyPair;
+            var issuerSerialNumber = serialNumber;
+            cg.AddExtension(X509Extensions.AuthorityKeyIdentifier.Id, false,
+                new AuthorityKeyIdentifier(SubjectPublicKeyInfoFactory.CreateSubjectPublicKeyInfo(subjectKeyPair.Public),
+                    new GeneralNames(new GeneralName(issuerDN)), issuerSerialNumber));
+
+            if (!isCA)
+            {
+                // Key usage 
+                cg.AddExtension(X509Extensions.KeyUsage, true,
+                    new KeyUsage(KeyUsage.DataEncipherment | KeyUsage.DigitalSignature |
+                        KeyUsage.NonRepudiation | KeyUsage.KeyCertSign | KeyUsage.KeyEncipherment));
+
+                // Extended Key usage
+                cg.AddExtension(X509Extensions.ExtendedKeyUsage, true,
+                    new ExtendedKeyUsage(new List<DerObjectIdentifier>() {
+                    new DerObjectIdentifier("1.3.6.1.5.5.7.3.1"), // server auth
+                    new DerObjectIdentifier("1.3.6.1.5.5.7.3.2"), // client auth
+                    }));
+
+                // subject alternate name
+                cg.AddExtension(X509Extensions.SubjectAlternativeName, false,
+                    new GeneralNames(new GeneralName[] {
+                    new GeneralName(GeneralName.UniformResourceIdentifier, applicationUri),
+                    new GeneralName(GeneralName.DnsName, domainNames[0])}));
+            }
+            else
+            {
+                // Key usage CA
+                cg.AddExtension(X509Extensions.KeyUsage, true,
+                    new KeyUsage(KeyUsage.CrlSign | KeyUsage.DigitalSignature | KeyUsage.KeyCertSign));
+            }
+
+            // sign certificate
+            ISignatureFactory signatureFactory = 
+                new Asn1SignatureFactory((hashSizeInBits < 256) ? "SHA1WITHRSA" : "SHA256WITHRSA", subjectKeyPair.Private, random);
+            Org.BouncyCastle.X509.X509Certificate x509 = cg.Generate(signatureFactory);
+
+            // create pkcs12 store for cert and private key
+            MemoryStream pfxData = new MemoryStream();
+            Pkcs12Store pkcsStore = new Pkcs12StoreBuilder().Build();
+            X509CertificateEntry[] chain = new X509CertificateEntry[1];
+            chain[0] = new X509CertificateEntry(x509);
+            pkcsStore.SetKeyEntry(applicationName, new AsymmetricKeyEntry(subjectKeyPair.Private), chain);
+            pkcsStore.Save(pfxData, null, random);
+
+            // merge into X509Certificate2
+            X509Certificate2 certificate = new X509Certificate2(pfxData.ToArray(), (string)null, X509KeyStorageFlags.Exportable | X509KeyStorageFlags.DefaultKeySet);
```

Comment:
Is this working on all platforms? I remember having issues with the BouncyCastle pfx format vs. what X509Certificate2 needs...